extensions/ql-vscode/src/remote-queries/variant-analysis-monitor.ts

```diff
@@ -0,0 +1,81 @@
+import * as vscode from 'vscode';
+import { Credentials } from '../authentication';
+import { Logger } from '../logging';
+import * as ghApiClient from './gh-api/gh-api-client';
+
+import { VariantAnalysis, VariantAnalysisStatus } from './shared/variant-analysis';
+import {
+  VariantAnalysis as VariantAnalysisApiResponse
+} from './gh-api/variant-analysis';
+import { VariantAnalysisMonitorResult } from './shared/variant-analysis-monitor-result';
+import { processFailureReason } from './variant-analysis-processor';
+
+export class VariantAnalysisMonitor {
+  // With a sleep of 5 seconds, the maximum number of attempts takes
+  // us to just over 2 days worth of monitoring.
+  public static maxAttemptCount = 17280;
+  public static sleepTime = 5000;
+
+  constructor(
+    private readonly extensionContext: vscode.ExtensionContext,
+    private readonly logger: Logger
+  ) {
+  }
+
+  public async monitorVariantAnalysis(
+    variantAnalysis: VariantAnalysis,
+    cancellationToken: vscode.CancellationToken
+  ): Promise<VariantAnalysisMonitorResult> {
+
+    const credentials = await Credentials.initialize(this.extensionContext);
+    if (!credentials) {
+      throw Error('Error authenticating with GitHub');
+    }
+
+    let variantAnalysisSummary: VariantAnalysisApiResponse;
+    let attemptCount = 0;
+    const scannedReposDownloaded: number[] = [];
+
+    while (attemptCount <= VariantAnalysisMonitor.maxAttemptCount) {
+      await this.sleep(VariantAnalysisMonitor.sleepTime);
+
+      if (cancellationToken && cancellationToken.isCancellationRequested) {
+        return { status: 'Cancelled', error: 'Variant Analysis was canceled.' };
+      }
+
+      variantAnalysisSummary = await ghApiClient.getVariantAnalysis(
+        credentials,
+        variantAnalysis.controllerRepoId,
+        variantAnalysis.id
+      );
+
+      if (variantAnalysisSummary.status == 'in_progress' && variantAnalysisSummary.failure_reason) {
```

Comment:
Is it possible to have `status: in_progress` but also have a `failure_reason`? Could you explain this bit here a little more?

Reply:
This was a leftover from an initial sketch where I was trying to map the states of the monitor to the states of the variant analysis ('In progress', 'CompletedSuccessfully', 'CompletedUnsuccessfully' etc). Discussed this with Robert and came to the conclusion that we don't actually want to have a one-to-one direct mapping.

This is a long way to say I'm removing the check for `status` and relying on the existence of a `failure_reason` to mark the variant analysis as failed. 

Later we'll flesh this out further by defining when to emit events to the UI based on the state of the variant analysis summary.

Reply:
Have made the change here: https://github.com/github/vscode-codeql/pull/1545/commits/e41b0ff779438d05e6cca0f179eefdf32e698215.